### PR TITLE
Add flash-attn

### PR DESCRIPTION
**Pull request by RaymondLi0**
Flash-attention, based on https://github.com/NVIDIA/Megatron-LM/pull/267
with support for MQA

### CODE REVIEW

**Comment by RaymondLi0** [comment]
Some tests with a 1B MQA model (santacoder's config: num_layers 24, num_heads 16, hidden_size 2048), bf16, on 1 A100 gpu.

With flash-attn, this model can be trained with sequences of length up to `8192`, and with full-recomputation up to `32768`.
With normal-attn, we only reach `2048`, or `8192` with selective or full recomputation.

Flash-attn is faster, especially for longer sequences: 
Time-per-iteration for seq-length `2048`: flash-attn: `19794.1` VS normal-attn: `22679.3`
Time-per-iteration for seq-length `4096`: flash-attn: `44040.8` VS normal-attn: `71740` (selective-recomputation)
Time-per-iteration for seq-length `8192`: flash-attn: `113715.5` VS normal-attn: `256122` (selective-recomputation)

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/raymond.li/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/raymond.li/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



use_flash_attn | seq_len | mbs | gbs | Activation-recomputation | mem_reserved (GB) | iteration_time | TFLOPs
-- | -- | -- | -- | -- | -- | -- | --
TRUE | 512 | 2 | 192 | None | 22.45 | 6323.2 | 109.18
TRUE | 1024 | 2 | 192 | None | 24.73 | 9888.1 | 145.64
TRUE | 2048 | 2 | 192 | None | 29.27 | 19794.1 | 157.51
TRUE | 4096 | 2 | 192 | None | 38.36 | 44040.8 | 163.15
TRUE | 8192 | 2 | 192 | None | 56.63 | 113715.5 | 159.75
TRUE | 16384 | 2 | 192 | None | OOM | OOM | OOM
TRUE | 512 | 2 | 192 | Full | 20.8 | 8606.4 | 104.65
TRUE | 1024 | 2 | 192 | Full | 21.4 | 12837.4 | 146.48
TRUE | 2048 | 2 | 192 | Full | 22.58 | 26172.6 | 155.8
TRUE | 4096 | 2 | 192 | Full | 25 | 58775.6 | 160.3
TRUE | 8192 | 2 | 192 | Full | 29.85 | 151869.1 | 157.44
TRUE | 16384 | 2 | 192 | Full | 36.95 | 442532.3 | 153.86
TRUE | 32768 | 2 | 192 | Full | 50.72 | 1440645.7 | 150.79
FALSE | 512 | 2 | 192 | None | 23.18 | 6138.9 | 110.4
FALSE | 1024 | 2 | 192 | None | 28.24 | 10404.8 | 137.88
FALSE | 2048 | 2 | 192 | None | 44.46 | 22679.3 | 137.47
FALSE | 4096 | 2 | 192 | None | OOM | OOM | OOM
FALSE | 512 | 2 | 192 | Selective | 22.18 | 6840.5 | 100.92
FALSE | 1024 | 2 | 192 | Selective | 24.18 | 11216.9 | 128.39
FALSE | 2048 | 2 | 192 | Selective | 28.67 | 25446.1 | 122.52
FALSE | 4096 | 2 | 192 | Selective | 38.95 | 71740 | 100.16
FALSE | 8192 | 2 | 192 | Selective | 65.74 | 256122 | 70.95
FALSE | 16384 | 2 | 192 | Selective | OOM | OOM | OOM
FALSE | 512 | 2 | 192 | Full | 20.8 | 8109.4 | 111.06
FALSE | 1024 | 2 | 192 | Full | 21.44 | 13569.5 | 138.58
FALSE | 2048 | 2 | 192 | Full | 23.3 | 29930.3 | 136.24
FALSE | 4096 | 2 | 192 | Full | 28.06 | 81514.7 | 115.58
FALSE | 8192 | 2 | 192 | Full | 46.4 | 276012.9 | 86.63
FALSE | 16384 | 2 | 192 | Full | OOM | OOM | OOM



</body>

</html>

**Comment by RaymondLi0** [comment]
Additional test currently running: training runs on 5k steps should give the same loss
normal-attn model VS flash-attn VS flash-attn with TP and SP

Screen Shot 2023-03-24 at 7 02 37 PM

File: megatron/model/transformer.py [hunk 1]
```diff
@@ -533,6 +589,25 @@ def __init__(self, init_method,
...
+                raise ImportError('FlashAttention is not installed, please install with '
+                                  'pip install flash-attn')
+            assert attention_type == AttnType.self_attn, ('FlashAttention code path only supports '
+                                                          'self-attention for now')
+            assert self.attn_mask_type == AttnMaskType.causal, ('FlashAttention code path only '
+                                                                'supports causal mask for now')
+            assert args.position_embedding_type != PositionEmbeddingType.alibi, \
+                ('FlashAttention does not support alibi positional embeddings yet')
+            if rearrange is None:
+                raise ImportError('einops is not installed, please install with pip install einops')
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:604]: Should it be added to the doc? Or better, it would be very easy to just use plain pytorch instead.

File: megatron/model/transformer.py [hunk 2]
```diff
@@ -533,6 +589,25 @@ def __init__(self, init_method,
...
+            
+            if self.checkpoint_core_attention:
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:606]: I think this should raise an error. Also, selective recomputation has a huge performance impact so we really want it. How hard would it be to implement for FlashAttention?

File: megatron/model/transformer.py [hunk 3]
```diff
@@ -699,13 +774,30 @@ def forward(self, hidden_states, attention_mask,
         # ==================================
         # core attention computation
         # ==================================
-
-        if self.checkpoint_core_attention:
-            context_layer = self._checkpointed_attention_forward(
-                query_layer, key_layer, value_layer, attention_mask, alibi)
+        if not self.use_flash_attn:
```
Comment (jlamypoirier) [inline megatron/model/transformer.py]: This would be clearer without the double negation.

File: megatron/model/transformer.py [hunk 4]
```diff
@@ -699,13 +774,30 @@ def forward(self, hidden_states, attention_mask,
...
+            
+            if self.attention_head_type == "multiquery":
+                sq, b, np, hn = query_layer.size()
+                # Expand kv to be compatible with flash-attn implementation
+                # [sq, b, 1, hn] -> [sq, b, np, hn]
+                key_layer = key_layer.expand((sq, b, np, hn))
+                value_layer = value_layer.expand((sq, b, np, hn))
+            q, k, v = [rearrange(x, 's b ... -> b s ...').contiguous()
+                       for x in (query_layer, key_layer, value_layer)]
+            if not self.sequence_parallel:
```
Comment (jlamypoirier) [inline megatron/model/transformer.py]: Again, double negation

File: megatron/model/transformer.py [hunk 5]
```diff
@@ -699,13 +774,30 @@ def forward(self, hidden_states, attention_mask,
...
-            context_layer = self.core_attention(
-                query_layer, key_layer, value_layer, attention_mask, alibi)
+            
+            if self.attention_head_type == "multiquery":
+                sq, b, np, hn = query_layer.size()
+                # Expand kv to be compatible with flash-attn implementation
+                # [sq, b, 1, hn] -> [sq, b, np, hn]
+                key_layer = key_layer.expand((sq, b, np, hn))
+                value_layer = value_layer.expand((sq, b, np, hn))
+            q, k, v = [rearrange(x, 's b ... -> b s ...').contiguous()
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:785]: That looks very bad. Megatron uses the `s b` format precisely to avoid this kind of reshape. If FlashAttention uses `b s` we should use that format instead. It should be OK to just comment the two conversions, at least without sequence parallelism (SP would need extra changes but we probably won't use it anyway) https://github.com/bigcode-project/Megatron-LM/blob/multi-query-attention/megatron/model/language_model.py#L240 https://github.com/bigcode-project/Megatron-LM/blob/multi-query-attention/megatron/model/gpt_model.py#L43

File: megatron/model/transformer.py [hunk 6]
```diff
@@ -699,13 +774,30 @@ def forward(self, hidden_states, attention_mask,
...
+                    query_layer, key_layer, value_layer, attention_mask, alibi)
         else:
-            context_layer = self.core_attention(
-                query_layer, key_layer, value_layer, attention_mask, alibi)
+            
+            if self.attention_head_type == "multiquery":
+                sq, b, np, hn = query_layer.size()
+                # Expand kv to be compatible with flash-attn implementation
+                # [sq, b, 1, hn] -> [sq, b, np, hn]
+                key_layer = key_layer.expand((sq, b, np, hn))
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:783]: I'm wondering if FlashAttention would work with just expand, that doesn't allocate new memory. If it were to work we would get the full benefits of FlashAttention for MQA. (I would expect it to enforce contiguous tensors but it's worth checking)

File: megatron/model/transformer.py [hunk 7]
```diff
@@ -699,13 +774,30 @@ def forward(self, hidden_states, attention_mask,
...
+                value_layer = value_layer.expand((sq, b, np, hn))
+            q, k, v = [rearrange(x, 's b ... -> b s ...').contiguous()
+                       for x in (query_layer, key_layer, value_layer)]
+            if not self.sequence_parallel:
+                with mpu.get_cuda_rng_tracker().fork():
```
Comment (jlamypoirier) [inline megatron/model/transformer.py:790]: I'm assuming it's the equivalent of this? https://github.com/bigcode-project/Megatron-LM/blob/multi-query-attention/megatron/model/transformer.py#L298

File: megatron/model/transformer.py (hunk 1 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py:604]: `einops` is a requirement that is also installed when you do `pip install flash-attn`

File: megatron/model/transformer.py (hunk 2 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py:606]: I'm not sure how hard it would be. Do you think applying the same `_checkpointed_attention_forward` code would work in this case?

File: megatron/model/transformer.py (hunk 3 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py]: Not really a "double" negation, but I see what you mean :p

File: megatron/model/transformer.py (hunk 5 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py:785]: Are suggesting to use `b s` through the whole transformer model?
I think that would require a big chunk of refactoring work, and also testing to make sure we are not breaking anything.
Looking at the nice performance improvements that flash-attn brings, I wouldn't take the risk of breaking everything else just to avoid a transpose here.

File: megatron/model/transformer.py (hunk 6 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py:783]: Are you asking whether if it would still work if we remove the call to `.contiguous()` on the next line?

File: megatron/model/transformer.py (hunk 7 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py:790]: I think so yes

File: megatron/model/transformer.py (hunk 2 above)
Comment (RaymondLi0) [reply megatron/model/transformer.py:606]: Also, do you think that even with flash-attention, selective recomputation would give a significant memory reduction compared to no-recomputation?

Comment (jlamypoirier) [reply megatron/model/transformer.py:606]: Hmm maybe you're right, the goal of selective recomputation is to avoid storing the part that's quadratic in sequence length, but FlashAttention is linear

Comment (jlamypoirier) [reply megatron/model/transformer.py:606]: Actually FlashAttention already implements recomputation, so no action needed.

File: megatron/model/transformer.py (hunk 3 above)
Comment (jlamypoirier) [reply megatron/model/transformer.py]: It is in the `else` 🤷‍♂️

File: megatron/model/transformer.py (hunk 5 above)
Comment (jlamypoirier) [reply megatron/model/transformer.py:785]: Actually the order only matters for attention (and sequence parallell), so it should just be about bypassing these two lines.

File: megatron/model/transformer.py (hunk 6 above)
Comment (jlamypoirier) [reply megatron/model/transformer.py:783]: That would almost certainly not work (transposed tensors are much harder to deal with), but maybe if we do the expand after the transpose or skip the transpose altogether.

File: megatron/model/transformer.py (hunk 5 above)
Comment (jlamypoirier) [reply megatron/model/transformer.py:785]: The transposes have a big impact on memory usage and a moderate one on speed (I think) so it's quite important.

**Review by jlamypoirier — APPROVED** [review]
The remaining comments on eliminating unnecessary ops are not essential and can be looked into later.

Great job!

File: megatron/model/transformer.py (hunk 2 above)
Comment (jlamypoirier) [reply megatron/model/transformer.py:606]: (So we can change back to warning)

Comment (RaymondLi0) [reply megatron/model/transformer.py:606]: Oh, nice! Then I'll change back to warning, yes
